src/util/schnorr.rs

```diff
@@ -76,9 +80,77 @@ impl TweakedPublicKey {
     pub fn new(key: PublicKey) -> TweakedPublicKey {
         TweakedPublicKey(key)
     }
-    
-    /// Returns the underlying public key 
+
+    /// Returns the underlying public key
     pub fn into_inner(self) -> PublicKey {
         self.0
     }
+}
+
+/// A BIP340-341 serialized schnorr signature with the corresponding hash type.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct SchnorrSig {
+    /// The underlying schnorr signature
+    pub sig: secp256k1::schnorrsig::Signature,
+    /// The corresponding hash type
+    pub hash_ty: SchnorrSigHashType,
+}
+
+impl SchnorrSig {
+
+    /// Deserialize from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, SchnorrSigError> {
+        if sl.len() == SCHNORRSIG_SIGNATURE_SIZE {
+            // default type
+            let sig = secp256k1::schnorrsig::Signature::from_slice(sl)?;
+            Ok( SchnorrSig { sig, hash_ty : SchnorrSigHashType::Default })
+        } else {
+            match sl.split_last() {
+                Some((hash_ty, sig)) => {
+                    let hash_ty = SchnorrSigHashType::from_u8(*hash_ty)
+                        .map_err(|_| SchnorrSigError::InvalidSighashType(*hash_ty))?;
+                    let sig = secp256k1::schnorrsig::Signature::from_slice(sig)?;
+                    Ok(SchnorrSig { sig, hash_ty })
+                },
+                None => {
+                    Err(SchnorrSigError::Secp256k1(secp256k1::Error::InvalidSignature))?
+                },
+            }
+        }
+    }
+
+    /// Serialize SchnorrSig
+    pub fn serialize(&self) -> Vec<u8> {
+        // TODO: add support to serialize to a writer to SerializedSig
+        let mut ser_sig = self.sig.as_ref().to_vec();
+        ser_sig.push(self.hash_ty as u8);
+        ser_sig
+    }
+}
+
+/// A schnorr sig related error.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum SchnorrSigError {
+    /// Base58 encoding error
+    InvalidSighashType(u8),
+    /// secp256k1-related error
+    Secp256k1(secp256k1::Error),
+}
+
+
+impl fmt::Display for SchnorrSigError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            SchnorrSigError::InvalidSighashType(hash_ty) =>
+                write!(f, "Invalid signature hash type {}", hash_ty),
+            SchnorrSigError::Secp256k1(ref e) => write!(f, "Key secp256k1 error: {}", e),
+        }
+    }
+}
+
+#[doc(hidden)]
```

Comment:
I'm not convinced this helps much. Most people would probably blindly try `?` and thus start implicitly relying on it.

Reply:
This is how it is done throughout the library. I never really thought if this was useful or not, but just have the hidden as a rule :( .